### PR TITLE
Upgrade to billing 3.0

### PR DESCRIPTION
**Pull request by britzl**

### CODE REVIEW

File: extension-iap/src/iap_android.cpp [hunk 1]
```diff
@@ -52,7 +52,12 @@ static IAP g_IAP;
 
 static int IAP_ProcessPendingTransactions(lua_State* L)
 {
-    //todo handle pending transactions if there is such thing on Android
+    IAP* iap = &g_IAP;
```
Comment (JCash) [inline extension-iap/src/iap_android.cpp:55]: unused

File: extension-iap/manifests/android/build.gradle [hunk 2]
```diff
@@ -0,0 +1,3 @@
+dependencies {
+    implementation 'com.android.billingclient:billing:3.0.0'
+}
```
Comment (JCash) [inline extension-iap/manifests/android/build.gradle:3]: I love the ease-of-use with these .gradle files now ^^

File: extension-iap/manifests/android/AndroidManifest.xml [hunk 3]
```diff
@@ -3,23 +3,11 @@
...
             <intent-filter>
                 <action android:name="com.amazon.inapp.purchasing.NOTIFY" android:permission="com.amazon.inapp.purchasing.Permission.NOTIFY" />
             </intent-filter>
         </receiver>
-
     </application>
-
-    <uses-permission android:name="com.android.vending.BILLING" />
-    
 </manifest>
```
Comment (JCash) [inline extension-iap/manifests/android/AndroidManifest.xml:13]: Less manifest junk, I love it!

File: extension-iap/src/java/com/defold/iap/IapGooglePlay.java [hunk 4]
```diff
@@ -5,507 +5,382 @@
...
-                    try {
-                        skuDetailsThread.join();
-                    } catch (InterruptedException e) {
-                        Log.wtf(TAG, "Failed to join thread", e);
-                    }
-                }
-            }
-        });
+        Log.d(TAG, "stop()");
+        billingClient.endConnection();
```
Comment (JCash) [inline extension-iap/src/java/com/defold/iap/IapGooglePlay.java:73]: Comparing with the old code, it was "secured" from being called from within a callback, which might come from another thread.

Is this no longer the case?

Comment (britzl) [reply extension-iap/src/java/com/defold/iap/IapGooglePlay.java:73]: I thought that wouldn't be a problem anymore, but the documentation isn't clear and their example is a bit bloated an uses a lifecycle manager instance: https://github.com/android/play-billing-samples/blob/master/ClassyTaxiJava/app/src/main/java/com/example/android/classytaxijava/billing/BillingClientLifecycle.java#L113

**Comment by AGulev** [comment]
I tried in one of my games and updated the plugin. When I launched the game for the first time I got a crash. But I can't reproduce it anymore.

```
AndroidRuntime  E  FATAL EXCEPTION: Thread-16453
                         E  Process: com.potatojam.onet_paradise.match_two_tiles, PID: 23300
                         E  java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Iterator java.util.List.iterator()' on a null object reference
                         E      at com.defold.iap.IapGooglePlay$5.onSkuDetailsResponse(IapGooglePlay.java:336)
                         E      at com.android.billingclient.api.BillingClientImpl.querySkuDetailsAsync(com.android.billingclient:billing@@3.0.0:259)
                         E      at com.defold.iap.IapGooglePlay.querySkuDetailsAsync(IapGooglePlay.java:348)
                         E      at com.defold.iap.IapGooglePlay.listItems(IapGooglePlay.java:366)
```

**Comment by britzl** [comment]
@AGulev I've added null checks for additional return values coming from Google Play

File: extension-iap/src/iap_android.cpp (hunk 1 above)
Comment (britzl) [reply extension-iap/src/iap_android.cpp]: Thanks! Fixed!

**Comment by britzl** [comment]
I'm going to go ahead and merge now and set this a pre-release and ask for community help to test.

**Pull request merged**